lightning/src/util/zbase32.rs

```diff
@@ -0,0 +1,137 @@
+// This is a modification of base32 encoding to support the zbase32 alphabet.
+// The original piece of software can be found at https://github.com/andreasots/base32
+
+/*
+Copyright (c) 2015 The base32 Developers
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+const ALPHABET: &'static [u8] = b"ybndrfg8ejkmcpqxot1uwisza345h769";
+
+pub fn encode(data: &[u8]) -> String {
+    let mut ret = Vec::with_capacity((data.len() + 3) / 4 * 5);
+
+    for chunk in data.chunks(5) {
+        let buf = {
+            let mut buf = [0u8; 5];
+            for (i, &b) in chunk.iter().enumerate() {
+                buf[i] = b;
+            }
+            buf
+        };
+
+        ret.push(ALPHABET[((buf[0] & 0xF8) >> 3) as usize]);
+        ret.push(ALPHABET[(((buf[0] & 0x07) << 2) | ((buf[1] & 0xC0) >> 6)) as usize]);
+        ret.push(ALPHABET[((buf[1] & 0x3E) >> 1) as usize]);
+        ret.push(ALPHABET[(((buf[1] & 0x01) << 4) | ((buf[2] & 0xF0) >> 4)) as usize]);
+        ret.push(ALPHABET[(((buf[2] & 0x0F) << 1) | (buf[3] >> 7)) as usize]);
+        ret.push(ALPHABET[((buf[3] & 0x7C) >> 2) as usize]);
+        ret.push(ALPHABET[(((buf[3] & 0x03) << 3) | ((buf[4] & 0xE0) >> 5)) as usize]);
+        ret.push(ALPHABET[(buf[4] & 0x1F) as usize]);
+    }
+
+    if data.len() % 5 != 0 {
```

Comment:
This if block can be more succinctly written as `ret.truncate((data.len() * 8 + 4) / 5);`